src/frontend/org/voltdb/calciteadapter/rel/physical/AbstractVoltDBPAggregate.java

```diff
@@ -0,0 +1,208 @@
+ * along with VoltDB.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.voltdb.calciteadapter.rel.physical;
+
+import java.util.List;
+
+import org.apache.calcite.plan.RelOptCluster;
+import org.apache.calcite.plan.RelOptCost;
+import org.apache.calcite.plan.RelOptPlanner;
+import org.apache.calcite.plan.RelTrait;
+import org.apache.calcite.plan.RelTraitSet;
+import org.apache.calcite.rel.RelCollation;
+import org.apache.calcite.rel.RelCollationTraitDef;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.RelWriter;
+import org.apache.calcite.rel.core.Aggregate;
+import org.apache.calcite.rel.core.AggregateCall;
+import org.apache.calcite.rel.metadata.RelMetadataQuery;
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.rel.type.RelDataTypeField;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.util.ImmutableBitSet;
+import org.voltdb.calciteadapter.converter.ExpressionTypeConverter;
+import org.voltdb.calciteadapter.converter.RelConverter;
+import org.voltdb.calciteadapter.converter.RexConverter;
+import org.voltdb.calciteadapter.planner.CalcitePlanningException;
+import org.voltdb.expressions.AbstractExpression;
+import org.voltdb.plannodes.AbstractPlanNode;
+import org.voltdb.plannodes.AggregatePlanNode;
+import org.voltdb.plannodes.NodeSchema;
+import org.voltdb.types.ExpressionType;
+
+public abstract class AbstractVoltDBPAggregate extends Aggregate implements VoltDBPRel {
+
+    // HAVING expression
+    final protected RexNode m_postPredicate;
+
+    /** Constructor */
+    protected AbstractVoltDBPAggregate(
+            RelOptCluster cluster,
+            RelTraitSet traitSet,
+            RelNode child,
+            boolean indicator,
+            ImmutableBitSet groupSet,
+            List<ImmutableBitSet> groupSets,
+            List<AggregateCall> aggCalls,
+            RexNode postPredicate) {
+      super(cluster, traitSet, child, indicator, groupSet, groupSets, aggCalls);
+      m_postPredicate = postPredicate;
+    }
+
+    @Override
+    public RelWriter explainTerms(RelWriter pw) {
+        super.explainTerms(pw);
+        if (m_postPredicate != null) {
+            pw.item("having", m_postPredicate);
+        }
+        return pw;
+    }
+
+    @Override
+    protected String computeDigest() {
+        String digest = super.computeDigest();
+        if (m_postPredicate != null) {
+            digest += m_postPredicate.toString();
+        }
+        return digest;
+    }
+
+    @Override
+    public RelOptCost computeSelfCost(RelOptPlanner planner,
+            RelMetadataQuery mq) {
+        double rowCount = getInput().estimateRowCount(mq);
+        RelTrait aggrCollationTrait = traitSet.getTrait(RelCollationTraitDef.INSTANCE);
+        // @TODO Give a discount to Serial Aggregation based on the numer of collation fields
+        if (aggrCollationTrait instanceof RelCollation) {
```

Comment:
I think it might be good to add comments explaining the discount.

Reply:
Added comments
        // Give a discount to the Aggregate based on the number of the collation fields.
        //  - Hash Aggregate - zero columns and zero discount
        //  - Serial Aggregate - the collation size is equal to the number of the GROUP BY columns
        //          and max discount 1 - 0.1 -  0.01 - 0.001 - ...
        //  - Partial Aggregate - anything in between
        // The required order will be enforced by some index which collation would match / satisfy
        // the aggregate's collation. If a table has more than one index multiple Aggregate / IndexScan
        // combinations are possible and we want to pick the one that has the maximum GROUP BY columns
        // covered resulting in a more efficient aggregation (less hashing)